# Sami Im

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('2015 - HW - 2.csv')

In [3]:
df

Country                       Region  Happiness Rank  \
0          Greece               Western Europe           102.0   
1            Laos            Southeastern Asia            99.0   
2        Slovenia   Central and Eastern Europe            55.0   
3          Zambia           Sub-Saharan Africa            85.0   
4     Philippines            Southeastern Asia            90.0   
..            ...                          ...             ...   
156   Netherlands               Western Europe             7.0   
157  Sierra Leone           Sub-Saharan Africa           123.0   
158        Mexico  Latin America and Caribbean            14.0   
159   South Korea                 Eastern Asia            47.0   
160          Mali           Sub-Saharan Africa           138.0   

     Happiness Score  Economy (GDP per Capita)    Family  \
0             48.570                  11.54060   9.29330   
1              4.876                   0.59066   0.73803   
2              5.848                   1.18498   1.27385   
3              5.129                   0.47038   0.91612   
4              5.073                   0.70532   1.03516   
..               ...                       ...       ...   
156           73.780                  13.29440  12.80170   
157            4.507                   0.33024   0.95571   
158            7.187                   1.02054   0.91451   
159            5.984                   1.24461   0.95774   
160            3.995                   0.26074   1.03526   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                     8.82130  0.76990                        0.13970   
1                     0.54909  0.59591                        0.24249   
2                     0.87337  0.60855                        0.03787   
3                     0.29924  0.48827                        0.12468   
4                     0.58114  0.62545                        0.12279   
..                        ...      ...                            ...   
156                   8.92840  6.15760                        3.18140   
157                   0.00000  0.40840                        0.08786   
158                   0.81444  0.48181                        0.21312   
159                   0.96538  0.33208                        0.07857   
160                   0.20583  0.38857                        0.12352   

    Generosity  Dystopia Residual  
0            0           18.01010  
1      0.42192            1.73799  
2      0.25328            1.61583  
3      0.19591            2.63430  
4      0.24991            1.75360  
..         ...                ...  
156      4.761           24.65700  
157    0.21488            2.51009  
158    0.14074            3.60214  
159    0.18557            2.21978  
160    0.18798            1.79293  

[161 rows x 11 columns]

<h1 style='color:#c842f5'> Steps to clean: </h1>
<br>
<b>

- Check if there is duplicate rows before doing anything and delete them.
- Check what are the datatypes for each column and see if its right logically
- Check for Inconsistency data and validate them.
- Check for missing values and find thier right or approx values.
- Detect outliers
    </b>

<h1 style="color: #34eb80">1. Check duplicates and delete them</h1>

In [4]:
df[df.duplicated()]

Country                           Region  Happiness Rank  Happiness Score  \
21  Jordan  Middle East and Northern Africa            82.0            5.192   
43  Angola               Sub-Saharan Africa           137.0            4.033   
66   Ghana               Sub-Saharan Africa           114.0            4.633   

    Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
21                   0.90198  1.05392                   0.69639  0.40661   
43                   0.75778  0.86040                   0.16683  0.10384   
66                   0.54558  0.67954                   0.40132  0.42342   

    Trust (Government Corruption) Generosity  Dystopia Residual  
21                        0.14293    0.11053            1.87996  
43                        0.07122    0.12344            1.94939  
66                        0.04355    0.23087            2.30919

In [5]:
df = df.drop_duplicates()

In [6]:
df[df.duplicated()]

Empty DataFrame
Columns: [Country, Region, Happiness Rank, Happiness Score, Economy (GDP per Capita), Family, Health (Life Expectancy), Freedom, Trust (Government Corruption), Generosity, Dystopia Residual]
Index: []

<h2 style="color:#3d6fad">We had 3 duplicate rows and we've deleted them </h2>
<br>

<h1 style="color: #34eb80">2. Check datatypes</h1>

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 158 entries, 0 to 160
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        158 non-null    object 
 1   Region                         158 non-null    object 
 2   Happiness Rank                 152 non-null    float64
 3   Happiness Score                152 non-null    float64
 4   Economy (GDP per Capita)       150 non-null    float64
 5   Family                         158 non-null    float64
 6   Health (Life Expectancy)       142 non-null    float64
 7   Freedom                        158 non-null    float64
 8   Trust (Government Corruption)  158 non-null    float64
 9   Generosity                     158 non-null    object 
 10  Dystopia Residual              158 non-null    float64
dtypes: float64(8), object(3)
memory usage: 14.8+ KB


<h2 style="color: #34b7eb">Conclusions:</h2>
<br>
<b>
    
- 158 rows and 11 columns
- 3 columns are objects rest are float
- Country and Region should be objects(String) but Generosity should be like others factors type = float
- Happiness Rank, Score, Econmy and Health columns have missing values
- Rank column can be int since its just ranking(leave it end while solve the missing values)
    </b>

In [8]:
df['Generosity'] = df['Generosity'].astype(float)

<h2 style="color: red"> We have a character ',' we need to change it to '.' </h2>
<br>

In [9]:
df['Generosity'] = df['Generosity'].str.replace(',', '.')
df['Generosity'] = df['Generosity'].astype(float)

C:\Users\hp\AppData\Local\Temp\ipykernel_20260\326580897.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Generosity'] = df['Generosity'].str.replace(',', '.')
C:\Users\hp\AppData\Local\Temp\ipykernel_20260\326580897.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Generosity'] = df['Generosity'].astype(float)


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 158 entries, 0 to 160
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        158 non-null    object 
 1   Region                         158 non-null    object 
 2   Happiness Rank                 152 non-null    float64
 3   Happiness Score                152 non-null    float64
 4   Economy (GDP per Capita)       150 non-null    float64
 5   Family                         158 non-null    float64
 6   Health (Life Expectancy)       142 non-null    float64
 7   Freedom                        158 non-null    float64
 8   Trust (Government Corruption)  158 non-null    float64
 9   Generosity                     158 non-null    float64
 10  Dystopia Residual              158 non-null    float64
dtypes: float64(9), object(2)
memory usage: 14.8+ KB


In [11]:
df.head(3)

Country                      Region  Happiness Rank  Happiness Score  \
0    Greece              Western Europe           102.0           48.570   
1      Laos           Southeastern Asia            99.0            4.876   
2  Slovenia  Central and Eastern Europe            55.0            5.848   

   Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                  11.54060  9.29330                   8.82130  0.76990   
1                   0.59066  0.73803                   0.54909  0.59591   
2                   1.18498  1.27385                   0.87337  0.60855   

   Trust (Government Corruption)  Generosity  Dystopia Residual  
0                        0.13970     0.00000           18.01010  
1                        0.24249     0.42192            1.73799  
2                        0.03787     0.25328            1.61583

<h2 style="color:#3d6fad"> After converting the type to float, we checked df.info() and displayed the data to make sure we are working correctly </h2>
<br>

<h1 style="color: #34eb80">3. Check for inconsistencies data</h1>

<h3> Since we have 9 columns float, they only have numeric values so we need to check String columns</h3>

In [12]:
df['Region'].value_counts()

Sub-Saharan Africa                 40
Central and Eastern Europe         29
Western Europe                     21
Latin America and Caribbean        21
Middle East and Northern Africa    21
Southeastern Asia                   9
Southern Asia                       7
Eastern Asia                        6
North America                       2
Australia and New Zealand           2
Name: Region, dtype: int64

## We need to display all of the countries to check them
<br>

In [13]:
df[df['Region'] == 'Southeastern Asia']

Country             Region  Happiness Rank  Happiness Score  \
1           Laos  Southeastern Asia            99.0            4.876   
4    Philippines  Southeastern Asia            90.0            5.073   
25      Cambodia  Southeastern Asia           145.0            3.819   
55       Myanmar  Southeastern Asia           129.0            4.307   
92     Indonesia  Southeastern Asia            74.0            5.399   
93      Thailand  Southeastern Asia            34.0            6.455   
120     Malaysia  Southeastern Asia            61.0            5.770   
130      Vietnam  Southeastern Asia            75.0            5.360   
147    Singapore  Southeastern Asia            24.0            6.798   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
1                     0.59066  0.73803                   0.54909  0.59591   
4                     0.70532  1.03516                   0.58114  0.62545   
25                    0.46038  0.62736                   0.61114  0.66246   
55                    0.27108  0.70905                   0.48246  0.44017   
92                    0.82827  1.08708                   0.63793  0.46611   
93                    0.96690  1.26504                   0.73850  0.55664   
120                   1.12486  1.07023                   0.72394  0.53024   
130                   0.63216  0.91226                   0.74676  0.59444   
147                   1.52186  1.02000                   1.02525  0.54252   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
1                          0.24249     0.42192            1.73799  
4                          0.12279     0.24991            1.75360  
25                         0.07247     0.40359            0.98195  
55                         0.19034     0.79588            1.41805  
92                         0.00000     0.51535            1.86399  
93                         0.03187     0.57630            2.31945  
120                        0.10501     0.33075            1.88541  
130                        0.10441     0.16860            2.20173  
147                        0.49210     0.31105            1.88501

In [14]:
df[df['Region'] == 'Southern Asia']

Country         Region  Happiness Rank  Happiness Score  \
17    Bangladesh  Southern Asia           109.0            4.694   
35     Sri Lanka  Southern Asia           132.0            4.271   
45         India  Southern Asia           117.0            4.565   
48   Afghanistan  Southern Asia           153.0            3.575   
60      Pakistan  Southern Asia            81.0            5.194   
139        Nepal  Southern Asia           121.0            4.514   
151       Bhutan  Southern Asia            79.0            5.253   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
17                    0.39753  0.43106                   0.60164  0.40820   
35                    0.83524  1.01905                   0.70806  0.53726   
45                    0.64499  0.38174                   0.51529  0.39786   
48                    0.31982  0.30285                   0.30335  0.23414   
60                    0.59543  0.41411                   0.51466  0.12102   
139                   0.35997  0.86449                   0.56874  0.38282   
151                   0.77042  1.10395                   0.57407  0.53206   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
17                         0.12569     0.21222            2.51767  
35                         0.09179     0.40828            0.67108  
45                         0.08492     0.26475            2.27513  
48                         0.09719     0.36510            1.95210  
60                         0.10464     0.33671            3.10709  
139                        0.05907     0.32296            1.95637  
151                        0.15445     0.47998            1.63794

In [15]:
df[df['Region'] == 'Eastern Asia']

Country        Region  Happiness Rank  Happiness Score  \
52         China  Eastern Asia            84.0            5.140   
53     Hong Kong  Eastern Asia            72.0            5.474   
117     Mongolia  Eastern Asia           100.0            4.874   
124       Taiwan  Eastern Asia            38.0            6.298   
125        Japan  Eastern Asia            46.0            5.987   
159  South Korea  Eastern Asia            47.0            5.984   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
52                    0.89012  0.94675                   0.81658  0.51697   
53                    1.38604  1.05818                   1.01328  0.59608   
117                   0.82819  1.30060                   0.60268  0.43626   
124                   1.29098  1.07617                   0.87530  0.39740   
125                   1.27074  1.25712                   0.99111  0.49615   
159                   1.24461  0.95774                   0.96538  0.33208   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
52                         0.02781     0.08185            1.86040  
53                         0.37124     0.39478            0.65429  
117                        0.02666     0.33230            1.34759  
124                        0.08129     0.25376            2.32323  
125                        0.18060     0.10705            1.68435  
159                        0.07857     0.18557            2.21978

In [16]:
df[df['Region'] =='North America']

Country         Region  Happiness Rank  Happiness Score  \
69   United States  North America             NaN            7.119   
146         Canada  North America             5.0            7.427   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
69                    1.39451  1.24711                   0.86179  0.54604   
146                   1.32629  1.32261                   0.90563  0.63297   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
69                         0.15890     0.40105            2.51011  
146                        0.32957     0.45811            2.45176

In [17]:
df[df['Region'] == 'Australia and New Zealand']

Country                     Region  Happiness Rank  Happiness Score  \
136    Australia  Australia and New Zealand            10.0            7.284   
149  New Zealand  Australia and New Zealand             9.0            7.286   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
136                   1.33358  1.30923                   0.93156  0.65124   
149                   1.25018  1.31967                   0.90837  0.63938   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
136                        0.35637     0.43562            2.26646  
149                        0.42922     0.47501            2.26425

In [18]:
df[df['Region'] == 'Middle East and Northern Africa']

Country                           Region  Happiness Rank  \
12                      Iraq  Middle East and Northern Africa           112.0   
15                     Qatar  Middle East and Northern Africa            28.0   
18   Palestinian Territories  Middle East and Northern Africa           108.0   
20                    Jordan  Middle East and Northern Africa            82.0   
27                     Egypt  Middle East and Northern Africa           135.0   
33              Saudi Arabia  Middle East and Northern Africa            35.0   
38                   Bahrain  Middle East and Northern Africa            49.0   
41      United Arab Emirates  Middle East and Northern Africa            20.0   
57                   Morocco  Middle East and Northern Africa           920.0   
59                    Kuwait  Middle East and Northern Africa            39.0   
76                     Yemen  Middle East and Northern Africa           136.0   
86                   Tunisia  Middle East and Northern Africa           107.0   
87                   Tunisia  Middle East and Northern Africa           107.0   
90                      Iran  Middle East and Northern Africa           110.0   
103                  Algeria  Middle East and Northern Africa            68.0   
108                    Libya  Middle East and Northern Africa            63.0   
118                  Lebanon  Middle East and Northern Africa           103.0   
131                     Oman  Middle East and Northern Africa            22.0   
133                   Israel  Middle East and Northern Africa            11.0   
141                    Syria  Middle East and Northern Africa           156.0   
155                   Turkey  Middle East and Northern Africa            76.0   

     Happiness Score  Economy (GDP per Capita)   Family  \
12             4.677                   0.98549  0.81889   
15             6.611                   1.69042  1.07860   
18             4.715                   0.59867  0.92558   
20             5.192                   0.90198  1.05392   
27             4.194                   0.88180  0.74700   
33             6.411                   1.39541  1.08393   
38             5.960                   1.32376  1.21624   
41             6.901                   1.42727  1.12575   
57             5.013                   0.73479  0.64095   
59             6.295                   1.55422  1.16594   
76             4.077                   0.54649  0.68093   
86             4.739                       NaN  0.60429   
87             4.739                   0.88113  0.60429   
90             4.686                   1.00880  0.54447   
103            5.605                   0.93929  1.07772   
108            5.754                   1.13145  1.11862   
118            4.839                   1.02564  0.80001   
131            6.853                   1.36011  1.08182   
133            7.278                   1.22857  1.22393   
141            3.006                   0.66320  0.47489   
155            5.332                   1.06098  0.94632   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
12                        NaN  0.00000                        0.13788   
15                    0.79733  0.64040                        0.52208   
18                    0.66015  0.24499                        0.12905   
20                    0.69639  0.40661                        0.14293   
27                    0.61712  0.17288                        0.06324   
33                    0.72025  0.31048                        0.32524   
38                    0.74716  0.45492                        0.30600   
41                    0.80925  0.64157                        0.38583   
57                    0.60954  0.41691                        0.08546   
59                    0.72492  0.55499                        0.25609   
76                    0.40064  0.35571                        0.07854   
86                    0.73793  0.26268                        0

<p style="color: purple; font-size: large"> We noticed in this region 'Middle East and Northern Africa' there is a duplicate for <b>Tunisia</b> country we didn't caught</p>
<br>

In [19]:
df[df['Country'] == 'Tunisia']

Country                           Region  Happiness Rank  Happiness Score  \
86  Tunisia  Middle East and Northern Africa           107.0            4.739   
87  Tunisia  Middle East and Northern Africa           107.0            4.739   

    Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
86                       NaN  0.60429                   0.73793  0.26268   
87                   0.88113  0.60429                   0.73793  0.26268   

    Trust (Government Corruption)  Generosity  Dystopia Residual  
86                        0.06358     0.06431            2.12466  
87                        0.06358     0.06431            2.12466

<p> It seems the first time it was inserted, they missed to fill the <b>Economy (GDP per Capita) column</b>, so they re-inserted it with the missing value<p/>

<h3 style="color:#3d6fad" > We don't need the first record, so we have to drop it</h3>

In [20]:
df = df.drop(86)

In [21]:
df[df['Country'] == 'Tunisia']

Country                           Region  Happiness Rank  Happiness Score  \
87  Tunisia  Middle East and Northern Africa           107.0            4.739   

    Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
87                   0.88113  0.60429                   0.73793  0.26268   

    Trust (Government Corruption)  Generosity  Dystopia Residual  
87                        0.06358     0.06431            2.12466

In [22]:
df[df['Region'] == 'Latin America and Caribbean']

Country                       Region  Happiness Rank  \
9            El Salvador  Latin America and Caribbean            42.0   
26                  Peru  Latin America and Caribbean            58.0   
34              Colombia  Latin America and Caribbean            33.0   
61             Argentina  Latin America and Caribbean            30.0   
62               Ecuador  Latin America and Caribbean            48.0   
67                 Chile  Latin America and Caribbean            27.0   
82                 Haiti  Latin America and Caribbean           119.0   
83               Uruguay  Latin America and Caribbean            32.0   
85             Nicaragua  Latin America and Caribbean            57.0   
100  Trinidad and Tobago  Latin America and Caribbean            41.0   
101              Jamaica  Latin America and Caribbean            65.0   
104           Costa Rica  Latin America and Caribbean            12.0   
110               Brazil  Latin America and Caribbean            16.0   
112             Paraguay  Latin America and Caribbean            53.0   
115               Panama  Latin America and Caribbean            25.0   
126            Guatemala  Latin America and Caribbean            43.0   
127             Honduras  Latin America and Caribbean           105.0   
137            Venezuela  Latin America and Caribbean            23.0   
140             Suriname  Latin America and Caribbean            40.0   
154              Bolivia  Latin America and Caribbean             NaN   
158               Mexico  Latin America and Caribbean            14.0   

     Happiness Score  Economy (GDP per Capita)   Family  \
9                NaN                   0.76454  1.02507   
26               NaN                       NaN  0.97459   
34             6.477                   0.91861  1.24018   
61             6.574                   1.05351  1.24823   
62             5.975                   0.86402  0.99903   
67             6.670                   1.10715  1.12447   
82             4.518                   0.26673  0.74302   
83             6.485                   1.06166  1.20890   
85             5.828                   0.59325  1.14184   
100            6.168                   1.21183  1.18354   
101            5.709                   0.81038  1.15102   
104              NaN                       NaN  1.23788   
110            6.983                   0.98124  1.23287   
112            5.878                   0.75985  1.30477   
115            6.786                   1.06353  1.19850   
126            6.123                   0.74553  1.04356   
127            4.788                   0.59532  0.95348   
137            6.810                   1.04424  1.25596   
140            6.269                   0.99534  0.97200   
154            5.890                   0.68133  0.97841   
158            7.187                   1.02054  0.91451   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
9                         NaN  0.40350                        0.11776   
26                        NaN  0.41496                        0.05989   
34                    0.69077  0.53466                        0.05120   
61                    0.78723  0.44974                        0.08484   
62                    0.79075  0.48574                        0.18090   
67                        NaN  0.44132                        0.12869   
82                    0.38847  0.24425                        0.17175   
83                        NaN  0.60362                        0.24558   
85                    0.74314  0.55475                        0.19317   
100                   0.61483  0.55884                        0.01140   
101                   0.68741  0.50442                        0.02299   
104                   0.86027  0.63376                        0.10583   
110                   0.69702  0.49049                        0.17521   
112                   0.66098  0.53899                        0.08242   
115             

In [23]:
df[df['Region'] == 'Sub-Saharan Africa']

Country              Region  Happiness Rank  \
3                      Zambia  Sub-Saharan Africa            85.0   
13                    Nigeria  Sub-Saharan Africa            78.0   
14                      Niger  Sub-Saharan Africa           144.0   
16                     Rwanda  Sub-Saharan Africa           154.0   
23                 Mozambique  Sub-Saharan Africa            94.0   
24               South Africa  Sub-Saharan Africa           113.0   
29        Congo (Brazzaville)  Sub-Saharan Africa           139.0   
32                 Madagascar  Sub-Saharan Africa           147.0   
36                      Kenya  Sub-Saharan Africa           125.0   
37                    Liberia  Sub-Saharan Africa           116.0   
39                     Guinea  Sub-Saharan Africa           150.0   
40                    Senegal  Sub-Saharan Africa           142.0   
42                     Angola  Sub-Saharan Africa           137.0   
46                      Sudan  Sub-Saharan Africa           118.0   
49                    Comoros  Sub-Saharan Africa           140.0   
50                   Cameroon  Sub-Saharan Africa           133.0   
54           Congo (Kinshasa)  Sub-Saharan Africa           120.0   
58                      Benin  Sub-Saharan Africa           155.0   
63                   Djibouti  Sub-Saharan Africa           126.0   
65                      Ghana  Sub-Saharan Africa           114.0   
70                  Mauritius  Sub-Saharan Africa            71.0   
71                    Burundi  Sub-Saharan Africa           157.0   
74                Ivory Coast  Sub-Saharan Africa           151.0   
89                   Zimbabwe  Sub-Saharan Africa           115.0   
91                   Ethiopia  Sub-Saharan Africa           122.0   
95               Burkina Faso  Sub-Saharan Africa           152.0   
99                     Malawi  Sub-Saharan Africa           131.0   
102                    Uganda  Sub-Saharan Africa           141.0   
106                      Chad  Sub-Saharan Africa           149.0   
107                   Lesotho  Sub-Saharan Africa            97.0   
109         Somaliland region  Sub-Saharan Africa            91.0   
119  Central African Republic  Sub-Saharan Africa           148.0   
135                  Tanzania  Sub-Saharan Africa           146.0   
142                  Botswana  Sub-Saharan Africa           128.0   
143                     Gabon  Sub-Saharan Africa           143.0   
148                 Swaziland  Sub-Saharan Africa           101.0   
150                Mauritania  Sub-Saharan Africa           124.0   
152                      Togo  Sub-Saharan Africa           158.0   
157              Sierra Leone  Sub-Saharan Africa           123.0   
160                      Mali  Sub-Saharan Africa           138.0   

     Happiness Score  Economy (GDP per Capita)   Family  \
3              5.129                   0.47038  0.91612   
13             5.268                   0.65435  0.90432   
14             3.845                       NaN  0.77265   
16             3.465                   0.22208  0.77370   
23             4.971                   0.08308  1.02626   
24             4.642                   0.92049  1.18468   
29             3.989                   0.67866  0.66290   
32             3.681                   0.20824  0.66801   
36             4.419                   0.36471  0.99876   
37             4.571                   0.07120  0.78968   
39             3.656                       NaN  0.46475   
40             3.904                   0.36498  0.97619   
42             4.033                   0.75778  0.86040   
46             4.550                   0.52107  1.01404   
49             3.956                   0.23906  0.79273   
50             4.252                   0.42250  0.88767   
54             4.517                   0.00000  1.00120   
58             3.340                   0.28665  0.35386   
63             4.369                   0.44025  0.59207   
65          

In [24]:
df[df['Region'] == 'Western Europe']

Country          Region  Happiness Rank  Happiness Score  \
0            Greece  Western Europe           102.0            48.57   
6             Italy  Western Europe            50.0            59.48   
10           France  Western Europe            29.0            65.75   
11            Malta  Western Europe            37.0            63.02   
19           Norway  Western Europe             4.0            75.22   
28          Iceland  Western Europe             2.0            75.61   
30            Spain  Western Europe             NaN            63.29   
51          Denmark  Western Europe             3.0            75.27   
68           Sweden  Western Europe             8.0            73.64   
77       Luxembourg  Western Europe            17.0            69.46   
79           Cyprus  Western Europe            67.0            56.89   
80          Belgium  Western Europe            19.0            69.37   
84          Ireland  Western Europe            18.0            69.40   
94      Switzerland  Western Europe             1.0            75.87   
105         Germany  Western Europe            26.0            67.50   
113         Austria  Western Europe            13.0            72.00   
121  United Kingdom  Western Europe            21.0              NaN   
128        Portugal  Western Europe            88.0            51.02   
132    North Cyprus  Western Europe            66.0            56.95   
153         Finland  Western Europe             6.0            74.06   
156     Netherlands  Western Europe             7.0            73.78   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                     11.5406   9.2933                    8.8213   0.7699   
6                     12.5114  11.9777                    9.5446   2.6236   
10                    12.7778  12.6038                    9.4579   5.5011   
11                    12.0740  13.0203                    8.8721   6.0365   
19                    14.5900  13.3095                       NaN   6.6973   
28                    13.0232  14.0223                    9.4784   6.2877   
30                    12.3011  13.1379                    9.5562   4.5951   
51                    13.2548  13.6058                    8.7464   6.4938   
68                    13.3171  12.8907                       NaN   6.5980   
77                    15.6391  12.1963                    9.1894   6.1583   
79                    12.0813   8.9318                    9.2356   4.0672   
80                    13.0782  12.8566                    8.9667   5.8450   
84                    13.3596  13.6948                    8.9533   6.1777   
94                    13.9651  13.4951                    9.4143   6.6557   
105                   13.2792  12.9937                    8.9186   6.1477   
113                   13.3723  12.9704                    8.9042   6.2433   
121                   12.6637  12.8548                       NaN   5.9625   
128                   11.5991  11.3935                    8.7519   5.1469   
132                   12.0806  10.7008                    9.2356   4.9027   
153                   12.9025  13.1826                    8.8911   6.4169   
156                   13.2944  12.8017                    8.9284   6.1576   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
0                           0.1397      0.0000            18.0101  
6                           0.2901      2.2823            20.2518  
10                          2.0646      1.2332            22.1126  
11                          1.3586      5.1752            16.4880  
19                          3.6503      3.4699            24.6531  
28                          1.4145      4.3630            27.0201  
30                          0.6398      1.8227            21.2367  
51                          4.8357      3.4139            24.9204  
68                          4.3844      3.6262            23.7119  
77                          3.7798      2.8034      

In [25]:
df[df['Region'] == 'Central and Eastern Europe']

Country                      Region  Happiness Rank  \
2                  Slovenia  Central and Eastern Europe            55.0   
5                   Estonia  Central and Eastern Europe            73.0   
7              Turkmenistan  Central and Eastern Europe            70.0   
8                   Georgia  Central and Eastern Europe           130.0   
22               Azerbaijan  Central and Eastern Europe            80.0   
31           Czech Republic  Central and Eastern Europe            31.0   
44                   Latvia  Central and Eastern Europe            89.0   
47   Bosnia and Herzegovina  Central and Eastern Europe            96.0   
56                 Slovakia  Central and Eastern Europe            45.0   
64                  Moldova  Central and Eastern Europe             NaN   
72               Montenegro  Central and Eastern Europe            82.0   
73                  Hungary  Central and Eastern Europe           104.0   
75                  Armenia  Central and Eastern Europe           127.0   
78                 Bulgaria  Central and Eastern Europe           134.0   
81                  Croatia  Central and Eastern Europe             NaN   
88                  Romania  Central and Eastern Europe             NaN   
96                Lithuania  Central and Eastern Europe            56.0   
97                  Ukraine  Central and Eastern Europe           111.0   
98               Uzbekistan  Central and Eastern Europe            44.0   
111                  Serbia  Central and Eastern Europe            87.0   
114                  Poland  Central and Eastern Europe            60.0   
116               Macedonia  Central and Eastern Europe            93.0   
122                 Belarus  Central and Eastern Europe            59.0   
123                 Albania  Central and Eastern Europe            95.0   
129                  Russia  Central and Eastern Europe            64.0   
134              Tajikistan  Central and Eastern Europe           106.0   
138              Kyrgyzstan  Central and Eastern Europe            77.0   
144              Kazakhstan  Central and Eastern Europe            54.0   
145                  Kosovo  Central and Eastern Europe            69.0   

     Happiness Score  Economy (GDP per Capita)   Family  \
2              5.848                   1.18498  1.27385   
5              5.429                   1.15174  1.22791   
7              5.548                   0.95847  1.22668   
8              4.297                   0.74190  0.38562   
22             5.212                   1.02389  0.93793   
31             6.505                   1.17898  1.20643   
44             5.098                   1.11312  1.09562   
47             4.949                   0.83223  0.91916   
56             5.995                   1.16891  1.26999   
64             5.889                   0.59448  1.01528   
72             5.192                   0.97438  0.90557   
73             4.800                   1.12094  1.20215   
75             4.350                   0.76821  0.77711   
78             4.218                   1.01216  1.10614   
81             5.759                   1.08254  0.79624   
88             5.124                   1.04345  0.88588   
96             5.833                   1.14723  1.25745   
97             4.681                   0.79907  1.20278   
98             6.003                   0.63244  1.34043   
111            5.123                   0.92053  1.00964   
114            5.791                   1.12555  1.27948   
116            5.007                   0.91851  1.00232   
122            5.813                   1.03192  1.23289   
123            4.959                   0.87867  0.80434   
129            5.716                   1.13764  1.23617   
134            4.786                   0.39047  0.85563   
138              NaN                       NaN  1.15115   
144            5.855                   1.12254  1.12241   
145            5.589                   0.80148  0.81198   

<h2 style="color:#3d6fad">I checked all countrie's names and letter cases, and all clear</h2>

<h1 style="color: #34eb80">4. Check missing values</h1>

In [26]:
df.isna().sum()

Country                           0
Region                            0
Happiness Rank                    6
Happiness Score                   6
Economy (GDP per Capita)          7
Family                            0
Health (Life Expectancy)         16
Freedom                           0
Trust (Government Corruption)     0
Generosity                        0
Dystopia Residual                 0
dtype: int64

### For Score column, we can get the value by summing all factors columns and get the score, if all factors are present
###  For Rank column, if we have the score value we can sort the df and see what rank it lays between

<h2> First get missing values from factors column </h2>

In [27]:
factors = {
    'eco': 'Economy (GDP per Capita)',
    'fam': 'Family',
    'health': 'Health (Life Expectancy)',
    'free': 'Freedom',
    'trust': 'Trust (Government Corruption)',
    'gen': 'Generosity',
    'dys': 'Dystopia Residual'
}

In [28]:
df[(df['Health (Life Expectancy)'].isna()) & (df['Happiness Score'].notna())].head(3)

Country                           Region  Happiness Rank  Happiness Score  \
12     Iraq  Middle East and Northern Africa           112.0            4.677   
13  Nigeria               Sub-Saharan Africa            78.0            5.268   
19   Norway                   Western Europe             4.0           75.220   

    Economy (GDP per Capita)    Family  Health (Life Expectancy)  Freedom  \
12                   0.98549   0.81889                       NaN  0.00000   
13                   0.65435   0.90432                       NaN  0.34334   
19                  14.59000  13.30950                       NaN  6.69730   

    Trust (Government Corruption)  Generosity  Dystopia Residual  
12                        0.13788     0.17922            1.95335  
13                        0.04030     0.27233            2.89319  
19                        3.65030     3.46990           24.65310

In [29]:
def calculate_missing(score, factors):
    '''
        This function takes the Happiness score value and the factors as list
        and returns the missing value
    '''
    return score - sum(factors)

In [30]:
df['Health (Life Expectancy)'] = df['Health (Life Expectancy)'].fillna(
    calculate_missing(df['Happiness Score'], [
        df[factors.get('eco')],
        df[factors.get('fam')],
        df[factors.get('free')],
        df[factors.get('trust')],
        df[factors.get('gen')],
        df[factors.get('dys')]
    ]))

In [31]:
df['Economy (GDP per Capita)'] = df['Economy (GDP per Capita)'].fillna(
    calculate_missing(df['Happiness Score'], [
        df[factors.get('health')],
        df[factors.get('fam')],
        df[factors.get('free')],
        df[factors.get('trust')],
        df[factors.get('gen')],
        df[factors.get('dys')]
    ]))

In [32]:
df['Happiness Score'] = df['Happiness Score'].fillna(sum([
    df[factors.get('eco')],    
    df[factors.get('fam')],
    df[factors.get('health')],
    df[factors.get('free')],
    df[factors.get('trust')],
    df[factors.get('gen')],
    df[factors.get('dys')]
    ])
) 

In [33]:
df.isna().sum()

Country                          0
Region                           0
Happiness Rank                   6
Happiness Score                  6
Economy (GDP per Capita)         4
Family                           0
Health (Life Expectancy)         5
Freedom                          0
Trust (Government Corruption)    0
Generosity                       0
Dystopia Residual                0
dtype: int64

<h2 style="color:#3d6fad"> Here we see that we decreased the number of missing values from Health & Econmy columns </h2>

<h1 style="color:red"> Notes: </h1>

<ul style="color: #3d6fad; font-weight: bold">
    <li>We have missing values in Happiness Rank and Happiness Score columns </li>
    <li>Happiness Rank is based on the Score column, so if we found the missing values in Happiness Score we can give each country its correct Rank. </li>
    <li> We can impute the missing Scores by setting the Region's mean, since all countries in the region have close scores value </li>
 </ul>

In [34]:
df[(df['Happiness Rank'].isna()) | (df['Happiness Score'].isna())]

Country                       Region  Happiness Rank  \
9       El Salvador  Latin America and Caribbean            42.0   
26             Peru  Latin America and Caribbean            58.0   
30            Spain               Western Europe             NaN   
64          Moldova   Central and Eastern Europe             NaN   
69    United States                North America             NaN   
81          Croatia   Central and Eastern Europe             NaN   
88          Romania   Central and Eastern Europe             NaN   
91         Ethiopia           Sub-Saharan Africa           122.0   
104      Costa Rica  Latin America and Caribbean            12.0   
121  United Kingdom               Western Europe            21.0   
138      Kyrgyzstan   Central and Eastern Europe            77.0   
154         Bolivia  Latin America and Caribbean             NaN   

     Happiness Score  Economy (GDP per Capita)    Family  \
9                NaN                   0.76454   1.02507   
26               NaN                       NaN   0.97459   
30            63.290                  12.30110  13.13790   
64             5.889                   0.59448   1.01528   
69             7.119                   1.39451   1.24711   
81             5.759                   1.08254   0.79624   
88             5.124                   1.04345   0.88588   
91               NaN                       NaN   0.60406   
104              NaN                       NaN   1.23788   
121              NaN                  12.66370  12.85480   
138              NaN                       NaN   1.15115   
154            5.890                   0.68133   0.97841   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
9                         NaN  0.40350                        0.11776   
26                        NaN  0.41496                        0.05989   
30                    9.55620  4.59510                        0.63980   
64                    0.61826  0.32818                        0.01615   
69                    0.86179  0.54604                        0.15890   
81                    0.78805  0.25883                        0.02430   
88                    0.76890  0.35068                        0.00649   
91                        NaN  0.43450                        0.15048   
104                   0.86027  0.63376                        0.10583   
121                       NaN  5.96250                        3.20670   
138                       NaN  0.43477                        0.04232   
154                   0.53920  0.57414                        0.08800   

     Generosity  Dystopia Residual  
9       0.10692            3.03500  
26      0.14982            2.59450  
30      1.82270           21.23670  
64      0.20951            3.10712  
69      0.40105            2.51011  
81      0.05444            2.75414  
88      0.13748            1.93129  
91      0.24325            2.44876  
104     0.25497            3.17728  
121     5.19120           19.69940  
138     0.30030            2.23270  
154     0.20536            2.82334

In [35]:
df.groupby('Region')['Happiness Score'].mean().to_dict()

{'Australia and New Zealand': 7.285,
 'Central and Eastern Europe': 5.334607142857143,
 'Eastern Asia': 5.626166666666666,
 'Latin America and Caribbean': 6.173222222222222,
 'Middle East and Northern Africa': 5.4069,
 'North America': 7.273,
 'Southeastern Asia': 5.317444444444444,
 'Southern Asia': 4.580857142857143,
 'Sub-Saharan Africa': 4.194871794871795,
 'Western Europe': 66.8075}

<h2> Above we can see the average for each region happiness score, we will need it to check if we filled the data correctly </h2>

In [36]:
regions = df['Region'].unique()
regions

array(['Western Europe', 'Southeastern Asia',
       'Central and Eastern Europe', 'Sub-Saharan Africa',
       'Latin America and Caribbean', 'Middle East and Northern Africa',
       'Southern Asia', 'Eastern Asia', 'North America',
       'Australia and New Zealand'], dtype=object)

In [37]:
df['Happiness Score'] = df.groupby('Region')['Happiness Score'].transform(lambda x: x.fillna(x.mean()))

In [38]:
df

Country                       Region  Happiness Rank  \
0          Greece               Western Europe           102.0   
1            Laos            Southeastern Asia            99.0   
2        Slovenia   Central and Eastern Europe            55.0   
3          Zambia           Sub-Saharan Africa            85.0   
4     Philippines            Southeastern Asia            90.0   
..            ...                          ...             ...   
156   Netherlands               Western Europe             7.0   
157  Sierra Leone           Sub-Saharan Africa           123.0   
158        Mexico  Latin America and Caribbean            14.0   
159   South Korea                 Eastern Asia            47.0   
160          Mali           Sub-Saharan Africa           138.0   

     Happiness Score  Economy (GDP per Capita)    Family  \
0             48.570                  11.54060   9.29330   
1              4.876                   0.59066   0.73803   
2              5.848                   1.18498   1.27385   
3              5.129                   0.47038   0.91612   
4              5.073                   0.70532   1.03516   
..               ...                       ...       ...   
156           73.780                  13.29440  12.80170   
157            4.507                   0.33024   0.95571   
158            7.187                   1.02054   0.91451   
159            5.984                   1.24461   0.95774   
160            3.995                   0.26074   1.03526   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                     8.82130  0.76990                        0.13970   
1                     0.54909  0.59591                        0.24249   
2                     0.87337  0.60855                        0.03787   
3                     0.29924  0.48827                        0.12468   
4                     0.58114  0.62545                        0.12279   
..                        ...      ...                            ...   
156                   8.92840  6.15760                        3.18140   
157                   0.00000  0.40840                        0.08786   
158                   0.81444  0.48181                        0.21312   
159                   0.96538  0.33208                        0.07857   
160                   0.20583  0.38857                        0.12352   

     Generosity  Dystopia Residual  
0       0.00000           18.01010  
1       0.42192            1.73799  
2       0.25328            1.61583  
3       0.19591            2.63430  
4       0.24991            1.75360  
..          ...                ...  
156     4.76100           24.65700  
157     0.21488            2.51009  
158     0.14074            3.60214  
159     0.18557            2.21978  
160     0.18798            1.79293  

[157 rows x 11 columns]

In [40]:
df[df['Happiness Score'].isna()]

Empty DataFrame
Columns: [Country, Region, Happiness Rank, Happiness Score, Economy (GDP per Capita), Family, Health (Life Expectancy), Freedom, Trust (Government Corruption), Generosity, Dystopia Residual]
Index: []

In [41]:
df.loc[9]

Country                                          El Salvador
Region                           Latin America and Caribbean
Happiness Rank                                          42.0
Happiness Score                                     6.173222
Economy (GDP per Capita)                             0.76454
Family                                               1.02507
Health (Life Expectancy)                                 NaN
Freedom                                               0.4035
Trust (Government Corruption)                        0.11776
Generosity                                           0.10692
Dystopia Residual                                      3.035
Name: 9, dtype: object

<H2 style="color:#3d6fad"> Here we can see we no longer have missing values in Happiness Score column and we checked one of the rows and its correct. </H2>

<h2> Now we can calculate the missing values in Econmy and health, we can just rerun the previous above cells that made this functionality </h2>

In [42]:
df.isna().sum()

Country                          0
Region                           0
Happiness Rank                   6
Happiness Score                  0
Economy (GDP per Capita)         4
Family                           0
Health (Life Expectancy)         5
Freedom                          0
Trust (Government Corruption)    0
Generosity                       0
Dystopia Residual                0
dtype: int64

In [43]:
df[df['Health (Life Expectancy)'].isna()]

Country                       Region  Happiness Rank  \
9       El Salvador  Latin America and Caribbean            42.0   
26             Peru  Latin America and Caribbean            58.0   
91         Ethiopia           Sub-Saharan Africa           122.0   
121  United Kingdom               Western Europe            21.0   
138      Kyrgyzstan   Central and Eastern Europe            77.0   

     Happiness Score  Economy (GDP per Capita)    Family  \
9           6.173222                   0.76454   1.02507   
26          6.173222                       NaN   0.97459   
91          4.194872                       NaN   0.60406   
121        66.807500                  12.66370  12.85480   
138         5.334607                       NaN   1.15115   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
9                         NaN  0.40350                        0.11776   
26                        NaN  0.41496                        0.05989   
91                        NaN  0.43450                        0.15048   
121                       NaN  5.96250                        3.20670   
138                       NaN  0.43477                        0.04232   

     Generosity  Dystopia Residual  
9       0.10692            3.03500  
26      0.14982            2.59450  
91      0.24325            2.44876  
121     5.19120           19.69940  
138     0.30030            2.23270

<h2> We only fixed 3 rows, and the others can't impute the missing values since we have two missing values on the same row </h2>
<br>

In [39]:
df[(df['Happiness Rank'].isnull())]

Country                       Region  Happiness Rank  \
30           Spain               Western Europe             NaN   
64         Moldova   Central and Eastern Europe             NaN   
69   United States                North America             NaN   
81         Croatia   Central and Eastern Europe             NaN   
88         Romania   Central and Eastern Europe             NaN   
154        Bolivia  Latin America and Caribbean             NaN   

     Happiness Score  Economy (GDP per Capita)    Family  \
30            63.290                  12.30110  13.13790   
64             5.889                   0.59448   1.01528   
69             7.119                   1.39451   1.24711   
81             5.759                   1.08254   0.79624   
88             5.124                   1.04345   0.88588   
154            5.890                   0.68133   0.97841   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
30                    9.55620  4.59510                        0.63980   
64                    0.61826  0.32818                        0.01615   
69                    0.86179  0.54604                        0.15890   
81                    0.78805  0.25883                        0.02430   
88                    0.76890  0.35068                        0.00649   
154                   0.53920  0.57414                        0.08800   

     Generosity  Dystopia Residual  
30      1.82270           21.23670  
64      0.20951            3.10712  
69      0.40105            2.51011  
81      0.05444            2.75414  
88      0.13748            1.93129  
154     0.20536            2.82334

<h2> Now we need to fix the rank, but first we will sort df based on score </h2>

In [44]:
df = df.sort_values(by='Happiness Score', ascending=False).reset_index(drop=True)
df.head(20)

Country          Region  Happiness Rank  Happiness Score  \
0      Switzerland  Western Europe             1.0          75.8700   
1          Iceland  Western Europe             2.0          75.6100   
2          Denmark  Western Europe             3.0          75.2700   
3           Norway  Western Europe             4.0          75.2200   
4          Finland  Western Europe             6.0          74.0600   
5      Netherlands  Western Europe             7.0          73.7800   
6           Sweden  Western Europe             8.0          73.6400   
7          Austria  Western Europe            13.0          72.0000   
8       Luxembourg  Western Europe            17.0          69.4600   
9          Ireland  Western Europe            18.0          69.4000   
10         Belgium  Western Europe            19.0          69.3700   
11         Germany  Western Europe            26.0          67.5000   
12  United Kingdom  Western Europe            21.0          66.8075   
13          France  Western Europe            29.0          65.7500   
14           Spain  Western Europe             NaN          63.2900   
15           Malta  Western Europe            37.0          63.0200   
16           Italy  Western Europe            50.0          59.4800   
17    North Cyprus  Western Europe            66.0          56.9500   
18          Cyprus  Western Europe            67.0          56.8900   
19        Portugal  Western Europe            88.0          51.0200   

    Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                    13.9651  13.4951                    9.4143   6.6557   
1                    13.0232  14.0223                    9.4784   6.2877   
2                    13.2548  13.6058                    8.7464   6.4938   
3                    14.5900  13.3095                    8.8499   6.6973   
4                    12.9025  13.1826                    8.8911   6.4169   
5                    13.2944  12.8017                    8.9284   6.1576   
6                    13.3171  12.8907                    9.1117   6.5980   
7                    13.3723  12.9704                    8.9042   6.2433   
8                    15.6391  12.1963                    9.1894   6.1583   
9                    13.3596  13.6948                    8.9533   6.1777   
10                   13.0782  12.8566                    8.9667   5.8450   
11                   13.2792  12.9937                    8.9186   6.1477   
12                   12.6637  12.8548                       NaN   5.9625   
13                   12.7778  12.6038                    9.4579   5.5011   
14                   12.3011  13.1379                    9.5562   4.5951   
15                   12.0740  13.0203                    8.8721   6.0365   
16                   12.5114  11.9777                    9.5446   2.6236   
17                   12.0806  10.7008                    9.2356   4.9027   
18                   12.0813   8.9318                    9.2356   4.0672   
19                   11.5991  11.3935                    8.7519   5.1469   

    Trust (Government Corruption)  Generosity  Dystopia Residual  
0                          4.1978      2.9678            25.1738  
1                          1.4145      4.3630            27.0201  
2                          4.8357      3.4139            24.9204  
3                          3.6503      3.4699            24.6531  
4                          4.1372      2.3351            26.1955  
5                          3.1814      4.7610            24.6570  
6                          4.3844      3.6262            23.7119  
7                          1.8676      3.3088            25.3320  
8                          3.7798      2.8034            19.6961  
9                          2.8703      4.5901            19.7570  
10                         2.2540      2.2250            24.1484  
11                         2.1843      2.8214            21.1569  
12                         3.2067      5.1912      

In [45]:
df.tail(20)

Country                           Region  \
137       Congo (Brazzaville)               Sub-Saharan Africa   
138                   Comoros               Sub-Saharan Africa   
139                    Uganda               Sub-Saharan Africa   
140                   Senegal               Sub-Saharan Africa   
141                     Gabon               Sub-Saharan Africa   
142                     Niger               Sub-Saharan Africa   
143                  Cambodia                Southeastern Asia   
144                  Tanzania               Sub-Saharan Africa   
145                Madagascar               Sub-Saharan Africa   
146  Central African Republic               Sub-Saharan Africa   
147                      Chad               Sub-Saharan Africa   
148                    Guinea               Sub-Saharan Africa   
149               Ivory Coast               Sub-Saharan Africa   
150              Burkina Faso               Sub-Saharan Africa   
151               Afghanistan                    Southern Asia   
152                    Rwanda               Sub-Saharan Africa   
153                     Benin               Sub-Saharan Africa   
154                     Syria  Middle East and Northern Africa   
155                   Burundi               Sub-Saharan Africa   
156                      Togo               Sub-Saharan Africa   

     Happiness Rank  Happiness Score  Economy (GDP per Capita)   Family  \
137           139.0            3.989                   0.67866  0.66290   
138           140.0            3.956                   0.23906  0.79273   
139           141.0            3.931                   0.21102  1.13299   
140           142.0            3.904                   0.36498  0.97619   
141           143.0            3.896                   1.06024  0.90528   
142           144.0            3.845                   0.06933  0.77265   
143           145.0            3.819                   0.46038  0.62736   
144           146.0            3.781                   0.28520  1.00268   
145           147.0            3.681                   0.20824  0.66801   
146           148.0            3.678                   0.07850  0.00000   
147           149.0            3.667                   0.34193  0.76062   
148           150.0            3.656                   0.17423  0.46475   
149           151.0            3.655                   0.46534  0.77115   
150           152.0            3.587                   0.25812  0.85188   
151           153.0            3.575                   0.31982  0.30285   
152           154.0            3.465                   0.22208  0.77370   
153           155.0            3.340                   0.28665  0.35386   
154           156.0            3.006                   0.66320  0.47489   
155           157.0            2.905                   0.01530  0.41587   
156           158.0            2.839                   0.20871  0.13995   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
137                   0.31051  0.41466                        0.11686   
138                   0.36351  0.22917                        0.19900   
139                   0.33861  0.45727                        0.07267   
140                   0.43540  0.36772                        0.10713   
141                   0.43372  0.31914                        0.11091   
142                   0.29707  0.47692                        0.15639   
143                   0.61114  0.66246                        0.07247   
144                   0.38215  0.32878                        0.05747   
145                   0.46721  0.19184                        0.08124   
146                   0.06699  0.48879                        0.08289   
147                   0.15010  0.23501                        0.05269   
148                   0.24009  0.37725                        0.12139   
149                   0.15185  0.46866                        0.17922   
150                   0.27125  0.3949

In [50]:
df['Happiness Rank'] = df.index + 1

In [58]:
df.head(20)

Country          Region  Happiness Rank  Happiness Score  \
0      Switzerland  Western Europe               1          75.8700   
1          Iceland  Western Europe               2          75.6100   
2          Denmark  Western Europe               3          75.2700   
3           Norway  Western Europe               4          75.2200   
4          Finland  Western Europe               5          74.0600   
5      Netherlands  Western Europe               6          73.7800   
6           Sweden  Western Europe               7          73.6400   
7          Austria  Western Europe               8          72.0000   
8       Luxembourg  Western Europe               9          69.4600   
9          Ireland  Western Europe              10          69.4000   
10         Belgium  Western Europe              11          69.3700   
11         Germany  Western Europe              12          67.5000   
12  United Kingdom  Western Europe              13          66.8075   
13          France  Western Europe              14          65.7500   
14           Spain  Western Europe              15          63.2900   
15           Malta  Western Europe              16          63.0200   
16           Italy  Western Europe              17          59.4800   
17    North Cyprus  Western Europe              18          56.9500   
18          Cyprus  Western Europe              19          56.8900   
19        Portugal  Western Europe              20          51.0200   

    Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                    13.9651  13.4951                    9.4143   6.6557   
1                    13.0232  14.0223                    9.4784   6.2877   
2                    13.2548  13.6058                    8.7464   6.4938   
3                    14.5900  13.3095                    8.8499   6.6973   
4                    12.9025  13.1826                    8.8911   6.4169   
5                    13.2944  12.8017                    8.9284   6.1576   
6                    13.3171  12.8907                    9.1117   6.5980   
7                    13.3723  12.9704                    8.9042   6.2433   
8                    15.6391  12.1963                    9.1894   6.1583   
9                    13.3596  13.6948                    8.9533   6.1777   
10                   13.0782  12.8566                    8.9667   5.8450   
11                   13.2792  12.9937                    8.9186   6.1477   
12                   12.6637  12.8548                       NaN   5.9625   
13                   12.7778  12.6038                    9.4579   5.5011   
14                   12.3011  13.1379                    9.5562   4.5951   
15                   12.0740  13.0203                    8.8721   6.0365   
16                   12.5114  11.9777                    9.5446   2.6236   
17                   12.0806  10.7008                    9.2356   4.9027   
18                   12.0813   8.9318                    9.2356   4.0672   
19                   11.5991  11.3935                    8.7519   5.1469   

    Trust (Government Corruption)  Generosity  Dystopia Residual  
0                          4.1978      2.9678            25.1738  
1                          1.4145      4.3630            27.0201  
2                          4.8357      3.4139            24.9204  
3                          3.6503      3.4699            24.6531  
4                          4.1372      2.3351            26.1955  
5                          3.1814      4.7610            24.6570  
6                          4.3844      3.6262            23.7119  
7                          1.8676      3.3088            25.3320  
8                          3.7798      2.8034            19.6961  
9                          2.8703      4.5901            19.7570  
10                         2.2540      2.2250            24.1484  
11                         2.1843      2.8214            21.1569  
12                         3.2067      5.1912      

# As we can see, we fixed the ranks values.
<br>


<h1 style="color: #34eb80">5. Detect outliers</h1>

In [115]:
# mean = df['Happiness Score'].mean()
# std = df['Happiness Score'].std()

# df['Z Score'] = (df['Happiness Score'] - mean) / std

In [116]:
# df

In [111]:
# outliers = df[(df['Happiness Score'] > 3 * std) | (df['Happiness Score'] < -3 * std)]

In [117]:
# outliers

In [73]:
df['Happiness Score'].describe()

count    157.000000
mean      13.414189
std       21.276105
min        2.839000
25%        4.518000
50%        5.332000
75%        6.455000
max       75.870000
Name: Happiness Score, dtype: float64

In [95]:
Q1 = 4.518000
Q3 = 6.455000
IQR = Q3 - Q1

In [96]:
above = Q3 + 1.5 * IQR
below = Q1 - 1.5 * IQR

In [98]:
outliers_2 = df[(df['Happiness Score'] > above) | (df['Happiness Score'] < below)]

In [99]:
outliers_2

Country          Region  Happiness Rank  Happiness Score  \
0      Switzerland  Western Europe               1          75.8700   
1          Iceland  Western Europe               2          75.6100   
2          Denmark  Western Europe               3          75.2700   
3           Norway  Western Europe               4          75.2200   
4          Finland  Western Europe               5          74.0600   
5      Netherlands  Western Europe               6          73.7800   
6           Sweden  Western Europe               7          73.6400   
7          Austria  Western Europe               8          72.0000   
8       Luxembourg  Western Europe               9          69.4600   
9          Ireland  Western Europe              10          69.4000   
10         Belgium  Western Europe              11          69.3700   
11         Germany  Western Europe              12          67.5000   
12  United Kingdom  Western Europe              13          66.8075   
13          France  Western Europe              14          65.7500   
14           Spain  Western Europe              15          63.2900   
15           Malta  Western Europe              16          63.0200   
16           Italy  Western Europe              17          59.4800   
17    North Cyprus  Western Europe              18          56.9500   
18          Cyprus  Western Europe              19          56.8900   
19        Portugal  Western Europe              20          51.0200   
20          Greece  Western Europe              21          48.5700   

    Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                    13.9651  13.4951                    9.4143   6.6557   
1                    13.0232  14.0223                    9.4784   6.2877   
2                    13.2548  13.6058                    8.7464   6.4938   
3                    14.5900  13.3095                    8.8499   6.6973   
4                    12.9025  13.1826                    8.8911   6.4169   
5                    13.2944  12.8017                    8.9284   6.1576   
6                    13.3171  12.8907                    9.1117   6.5980   
7                    13.3723  12.9704                    8.9042   6.2433   
8                    15.6391  12.1963                    9.1894   6.1583   
9                    13.3596  13.6948                    8.9533   6.1777   
10                   13.0782  12.8566                    8.9667   5.8450   
11                   13.2792  12.9937                    8.9186   6.1477   
12                   12.6637  12.8548                       NaN   5.9625   
13                   12.7778  12.6038                    9.4579   5.5011   
14                   12.3011  13.1379                    9.5562   4.5951   
15                   12.0740  13.0203                    8.8721   6.0365   
16                   12.5114  11.9777                    9.5446   2.6236   
17                   12.0806  10.7008                    9.2356   4.9027   
18                   12.0813   8.9318                    9.2356   4.0672   
19                   11.5991  11.3935                    8.7519   5.1469   
20                   11.5406   9.2933                    8.8213   0.7699   

    Trust (Government Corruption)  Generosity  Dystopia Residual   Z Score  
0                          4.1978      2.9678            25.1738  2.935491  
1                          1.4145      4.3630            27.0201  2.923271  
2                          4.8357      3.4139            24.9204  2.907290  
3                          3.6503      3.4699            24.6531  2.904940  
4                          4.1372      2.3351            26.1955  2.850419  
5                          3.1814      4.7610            24.6570  2.837259  
6                          4.3844      3.6262            23.7119  2.830678  
7                          1.8676      3.3088            25.3320  2.753597  
8                          3.7798      2.8034            19.6961  2.634214  
9    

<h1> I used the quantile method to detect outliers and we can notice all of them are from western europe </h1>
    
<h1 style='color: #5cd66e'> The dataframe is ready to use
</h1>